java/java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToRecordPattern.java

```diff
@@ -0,0 +1,190 @@
+import com.sun.source.util.TreePathScanner;
+import com.sun.source.tree.VariableTree;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.RecordComponentElement;
+import javax.lang.model.element.TypeElement;
+import org.netbeans.api.java.source.CodeStyleUtils;
+import org.netbeans.api.java.source.CompilationInfo;
+import org.netbeans.api.java.source.TreePathHandle;
+import org.netbeans.api.java.source.ElementHandle;
+import org.netbeans.api.java.source.WorkingCopy;
+import org.netbeans.api.java.queries.CompilerOptionsQuery;
+import org.netbeans.modules.java.hints.errors.Utilities;
+import org.netbeans.spi.editor.hints.ErrorDescription;
+import org.netbeans.spi.editor.hints.Fix;
+import org.netbeans.spi.java.hints.ErrorDescriptionFactory;
+import org.netbeans.spi.java.hints.Hint;
+import org.netbeans.spi.java.hints.HintContext;
+import org.netbeans.spi.java.hints.JavaFix;
+import org.netbeans.spi.java.hints.MatcherUtilities;
+import org.netbeans.spi.java.hints.TriggerPattern;
+import org.netbeans.spi.java.hints.TriggerPatterns;
+import org.openide.util.NbBundle;
+
+/**
+ *
+ * @author mjayan
+ */
+@NbBundle.Messages({
+    "DN_ConvertToRecordPattern=Convert to instanceof <record pattern>",
+    "DESC_ConvertToRecordPattern=Convert to instanceof <record pattern>",
+    "ERR_ConvertToRecordPattern=instanceof <record pattern> can be used here",
+    "FIX_ConvertToRecordPattern=Use instanceof <record pattern>"
+})
+@Hint(displayName = "#DN_ConvertToRecordPattern", description = "#DESC_ConvertToRecordPattern", category = "rules15",
+        minSourceVersion = "19")
+public class ConvertToRecordPattern {
+ 
+    private static final int RECORD_PATTERN_PREVIEW_JDK_VERSION = 19;
+
+    @TriggerPatterns({
+        @TriggerPattern(value = "if ($expr instanceof $typeI0 $var0 ) { $statements$;} else $else$;")
+    })
+    public static ErrorDescription trivial(HintContext ctx) {
+        if (Utilities.isJDKVersionLower(RECORD_PATTERN_PREVIEW_JDK_VERSION) && !CompilerOptionsQuery.getOptions(ctx.getInfo().getFileObject()).getArguments().contains("--enable-preview")) {
+            return null;
+        }
+        ElementKind kind = ctx.getInfo().getTrees().getElement(ctx.getVariables().get("$typeI0")).getKind();
+        if (kind.equals(ElementKind.RECORD)) {
+            Set<TreePath> convertPath = new HashSet<>();
+            Set<String> localVarList = new HashSet<>();
+            localVarList.add(ctx.getInfo().getTrees().getElement(ctx.getVariables().get("$expr")).getSimpleName().toString());
+            Map<String, String> varNames = new HashMap<>();
+            new TreePathScanner<Void, Void>() {
```

Comment:
Nit: possibly `CancelableTreePathScanner` (and use `HintContext.isCanceled()` to cancel).